File: integrations/mem0/python/src/purra_mem0/relations.py
"""Read-only, host-directed relation proposals over exact memory revisions."""
from collections.abc import Awaitable, Callable, Mapping, Sequence
from dataclasses import dataclass
from types import MappingProxyType

from ._journal import MemoryError
from .memory import Mem0Memory, MemoryRef, MemorySource


@dataclass(frozen=True, slots=True)
class MemoryRelationProposal:
    from_ref: MemoryRef
    to_ref: MemoryRef
    relation: str
    from_quote: str
    to_quote: str
    from_source: MemorySource
    to_source: MemorySource


async def propose_memory_relations(
    memory: Mem0Memory, refs: Sequence[MemoryRef], *, relations: Sequence[str],
    extract: Callable[[Mapping, object], Awaitable[object]],
    max_input_chars: int = 32_000, max_proposals: int = 32, signal=None,
) -> tuple[MemoryRelationProposal, ...]:
    """Call the host extractor once without linking records.

    The host owns Provider authorization, timeouts, budgets and semantic review.
    Quotes establish source membership, not the truth of an inferred relation.
    """
    if not callable(extract):
        raise TypeError("extract must be an async callable")
    for value, limit in ((max_input_chars, 10_000_000), (max_proposals, 256)):
        if type(value) is not int or not 1 <= value <= limit:
            raise ValueError("invalid relation extraction limit")
    if isinstance(refs, (str, bytes)) or not isinstance(refs, Sequence):
        raise TypeError("refs must be a sequence")
    refs = tuple(refs)
    if not 2 <= len(refs) <= 32 or any(not isinstance(ref, MemoryRef) for ref in refs):
        raise ValueError("provide 2 to 32 memory references")
    if len({ref.id for ref in refs}) != len(refs):
        raise ValueError("duplicate memory reference")
    if isinstance(relations, (str, bytes)) or not isinstance(relations, Sequence):
        raise TypeError("relations must be a sequence")
    relations = tuple(relations)
    if not 1 <= len(relations) <= 64 or any(
        not isinstance(value, str) or not value.strip() or value != value.strip() or len(value) > 64
        for value in relations
    ) or len(set(relations)) != len(relations):
        raise ValueError("invalid host relation vocabulary")
    epoch = memory.epoch

    def check():
        if signal is not None and signal.is_set():
            raise MemoryError("memory_cancelled")
        memory.assert_epoch(epoch)

    check()
    records = {}
    for ref in refs:
        record = await memory.get(ref.id, signal=signal)
        if record is None or record.version != ref.version:
            raise MemoryError("memory_relation_stale")
        records[ref.id] = record
    if sum(len(record.text) for record in records.values()) > max_input_chars:
        raise ValueError("relation source text exceeds max_input_chars")
    check()
    payload = MappingProxyType({
        "relations": relations, "maxProposals": max_proposals,
        "records": tuple(MappingProxyType({"id": r.id, "version": r.version, "text": r.text}) for r in records.values()),
    })
    raw = await extract(payload, signal)
    check()
    if not isinstance(raw, (list, tuple)) or len(raw) > max_proposals:
        raise MemoryError("memory_invalid_relation_proposal")
    result, seen = [], set()
    for item in raw:
        if not isinstance(item, Mapping) or set(item) != {"from", "to", "relation", "fromQuote", "toQuote"}:
            raise MemoryError("memory_invalid_relation_proposal")
        if any(not isinstance(value, str) for value in item.values()):
            raise MemoryError("memory_invalid_relation_proposal")
        a, b = records.get(item["from"]), records.get(item["to"])
        identity = (item["from"], item["to"], item["relation"])
        if (a is None or b is None or a.id == b.id or item["relation"] not in relations
                or identity in seen or any(not item[key].strip() or item[key] not in record.text
                    for key, record in (("fromQuote", a), ("toQuote", b)))):
            raise MemoryError("memory_invalid_relation_proposal")
        seen.add(identity)
        result.append(MemoryRelationProposal(MemoryRef(a.id, a.version), MemoryRef(b.id, b.version),
            item["relation"], item["fromQuote"], item["toQuote"], a.source, b.source))
    for ref in refs:
        if await memory.get(ref.id, signal=signal) != records[ref.id]:
            raise MemoryError("memory_relation_stale")
    check()
    return tuple(result)

File: integrations/mem0/typescript/src/relations.ts
import { MemoryError } from './journal.js';
import type { Mem0Memory, MemoryRef, MemorySource } from './memory.js';

export interface MemoryRelationProposal {
  readonly from: MemoryRef; readonly to: MemoryRef; readonly relation: string;
  readonly fromQuote: string; readonly toQuote: string;
  readonly fromSource: MemorySource; readonly toSource: MemorySource;
}
export interface MemoryRelationExtractionInput {
  readonly relations: readonly string[]; readonly maxProposals: number;
  readonly records: readonly { readonly id: string; readonly version: number; readonly text: string }[];
}
export type MemoryRelationExtractor = (input: MemoryRelationExtractionInput, signal?: AbortSignal) => Promise<unknown>;

/** Returns candidates only. The host owns model budgets, authorization and review. */
export async function proposeMemoryRelations(memory: Mem0Memory, refs: readonly MemoryRef[], options: {
  relations: readonly string[]; extract: MemoryRelationExtractor;
  maxInputChars?: number; maxProposals?: number; signal?: AbortSignal;
}): Promise<readonly MemoryRelationProposal[]> {
  const { extract, signal } = options;
  const maxInputChars = options.maxInputChars ?? 32_000, maxProposals = options.maxProposals ?? 32;
  if (typeof extract !== 'function') throw new TypeError('extract must be an async function');
  for (const [value, limit] of [[maxInputChars, 10_000_000], [maxProposals, 256]] as const) {
    if (!Number.isSafeInteger(value) || value < 1 || value > limit) throw new TypeError('Invalid relation extraction limit');
  }
  if (!Array.isArray(refs) || refs.length < 2 || refs.length > 32 || refs.some(ref =>
    !ref || typeof ref.id !== 'string' || !ref.id.trim() || ref.id.length > 512 ||
    !Number.isSafeInteger(ref.version) || ref.version < 1 || ref.version > 2 ** 31 - 2)) throw new TypeError('Provide 2 to 32 memory references');
  const selected = refs.map(ref => Object.freeze({ id: ref.id, version: ref.version }));
  if (new Set(selected.map(ref => ref.id)).size !== selected.length) throw new TypeError('Duplicate memory reference');
  if (!Array.isArray(options.relations) || options.relations.length < 1 || options.relations.length > 64 || options.relations.some(value =>
    typeof value !== 'string' || !value.trim() || value !== value.trim() || [...value].length > 64)) throw new TypeError('Invalid host relation vocabulary');
  const relations = Object.freeze([...options.relations]);
  if (new Set(relations).size !== relations.length) throw new TypeError('Duplicate relation');
  const epoch = memory.epoch;
  const check = () => {
    if (signal?.aborted) throw new MemoryError('memory_cancelled');
    memory.assertEpoch(epoch);
  };
  check();
  const records = new Map<string, NonNullable<Awaited<ReturnType<Mem0Memory['get']>>>>();
  for (const ref of selected) {
    const record = await memory.get(ref.id, signal === undefined ? {} : { signal });
    if (!record || record.version !== ref.version) throw new MemoryError('memory_relation_stale');
    records.set(ref.id, record);
  }
  if ([...records.values()].reduce((sum, record) => sum + [...record.text].length, 0) > maxInputChars) throw new TypeError('Relation source text exceeds maxInputChars');
  check();
  const raw = await extract(Object.freeze({ relations, maxProposals,
    records: Object.freeze([...records.values()].map(({ id, version, text }) => Object.freeze({ id, version, text }))),
  }), signal);
  check();
  if (!Array.isArray(raw) || raw.length > maxProposals) throw new MemoryError('memory_invalid_relation_proposal');
  const result: MemoryRelationProposal[] = [], seen = new Set<string>();
  for (const item of raw) {
    const keys = ['from', 'to', 'relation', 'fromQuote', 'toQuote'];
    if (!item || typeof item !== 'object' || Object.keys(item).length !== keys.length ||
      keys.some(key => !Object.hasOwn(item, key) || typeof item[key] !== 'string')) throw new MemoryError('memory_invalid_relation_proposal');
    const a = records.get(item.from), b = records.get(item.to);
    const identity = JSON.stringify([item.from, item.to, item.relation]);
    if (!a || !b || a.id === b.id || !relations.includes(item.relation) || seen.has(identity) ||
      !item.fromQuote.trim() || !a.text.includes(item.fromQuote) || !item.toQuote.trim() || !b.text.includes(item.toQuote)) throw new MemoryError('memory_invalid_relation_proposal');
    seen.add(identity);
    result.push(Object.freeze({ from: Object.freeze({ id: a.id, version: a.version }), to: Object.freeze({ id: b.id, version: b.version }),
      relation: item.relation, fromQuote: item.fromQuote, toQuote: item.toQuote,
      fromSource: Object.freeze({ ...a.source }), toSource: Object.freeze({ ...b.source }) }));
  }
  for (const ref of selected) {
    const before = records.get(ref.id)!, after = await memory.get(ref.id, signal === undefined ? {} : { signal });
    if (!after || after.version !== before.version || after.text !== before.text || after.state !== before.state ||
      after.source.id !== before.source.id || after.source.revision !== before.source.revision) throw new MemoryError('memory_relation_stale');
  }
  check();
  return Object.freeze(result);
}
